Objects/object.c

```diff
@@ -1577,9 +1579,11 @@ _PyObject_GenericSetAttrWithDict(PyObject *obj, PyObject *name,
         }
         else {
             PyErr_Format(PyExc_AttributeError,
-                         "'%.100s' object has no attribute '%U'",
+                         "'%.100s' object has no attribute '%U' and no "
+                         "__dict__ for setting new attributes",
```

Comment:
This path *may* have `__dict__` available. The only example I can think of that gets to this path with `__dict__` available is this:
```py
import _thread
a = _thread._local()
print(a.__dict__) # {}
del a.b # AttributeError!
```
(from [`local_setattro` in _threadmodule.c](https://github.com/python/cpython/blob/main/Modules/_threadmodule.c#L954))

Reply:
Why would that case go through setattr though?

Reply:
> Why would that case go through setattr though?

@Gobot1234 `setattro` handles both deletion (`value` is `NULL`) and setting of attributes (`value` is an object). It's also handled as a case in [`_PyObject_GenericSetAttrWithDict`](https://github.com/python/cpython/blob/main/Objects/object.c#L1565-L1568):
```c
if (value == NULL)
    res = PyDict_DelItem(dict, name);
else
    res = PyDict_SetItem(dict, name, value);
```